### PR TITLE
Jsonized monster attacks

### PR DESCRIPTION
**Pull request by Coolthulhu**
Similar to `iuse_actor`, except for monster attacks.

Allows easily defining per-monster-type attack variations, like for example leap range, bite damage or move cost of executing any attack.

Changed two monsters as examples:
- Feral predator will jump 8 tiles, but only if it has a target and said target is within 7 tiles. This is also a bit of a balance change: survivor can now sprint away from an angry predator, provided they start running before it closes in to 7 tiles of distance. Cooldown increased to 10 turns.
- Giant frog will leap for free every 10 turns. This means it will leap and then immediately move (possibly attack). It is still quite slow, so it doesn't pose any danger.

### CODE REVIEW

**Comment by SeanMirrsen** [comment]
This may be extraneous in principle, but maybe add a field for custom message string? On the off chance various attacks can be added with same base effect but different parameters, for variety purposes.

**Comment by Coolthulhu** [comment]
The leap is just an example. Making it print "hops" vs. "leaps" wouldn't change much.

**Comment by SeanMirrsen** [comment]
That's why I said "for variety purposes". It would only change the message string, but it might mean the difference between the message being seen as a warning ("The cougar leaps!") or as a harmless background event ("The tree frog hops around.")

**Comment by Coolthulhu** [comment]
Non-hostile critters don't leap around at the moment. When they start, it will be easy to add anyway.

**Comment by mugling** [comment]
@Coolthulhu This looks like a promising idea and hopefully a lot of interesting mechanics can be implemented following this PR.

> Variety is the spice of life, in wide open sandbox games especially.

Except if your the programmer (more complexity) or especially the translator (more workload).

> might mean the difference between the message being seen as a warning ("The cougar leaps!") or as a harmless background event ("The tree frog hops around.")

And in French, Italian or Korean? There's a lot that can be lost in translation.

> Making it print "hops" vs. "leaps" wouldn't change much.

I would agree. For any actor there should be one carefully chosen message. Also helpful for new players learning the game.

**Comment by Coolthulhu** [comment]
> For any actor there should be one carefully chosen message.

Well, some of the leaps can be quite different (for example, you could write a "travel leap" and "ambush leap"), it's just that in this example, all leapers are hostile critters who are highly likely to use the leap to attack something.

**Comment by mugling** [comment]
So in that example we could (later) consider adding an `is_aggressive` or `is_dangerous` field and show a differing message dependent upon that.

In general specifying the differences then deriving messages from them would seem preferable to allowing free form text for the reasons above. It also provides a context for further extension - perhaps your `TRIGGERHAPPY` character should get nervous and fire uncontrollably at a feral predator making an ambush (`is_aggressive`) leap but certainly for a leaping frog?

If a format string is an option it is less likely any future authors will define the differing aspects and specify them as parameters so we lose the possibility of such emergent behaviours.

File: src/mattack_actors.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#ifndef MATTACK_ACTORS_H
+#define MATTACK_ACTORS_H
+
+#include "json.h"
+#include "monattack.h"
```
Comment (BevapDin) [inline src/mattack_actors.h]: json.h doesn't need to be included here, forward declaring JsonObject should be enough (it's needed in the cpp file). Similar: monattack.h doesn't seem to be required either.

File: src/monster.h [hunk 2]
```diff
@@ -407,7 +407,7 @@ class monster : public Creature, public JsonSerializer, public JsonDeserializer
 
     private:
         int hp;
-        std::unordered_map<std::string, mon_special_attack> special_attacks;
+        std::map<std::string, mon_special_attack> special_attacks;
```
Comment (BevapDin) [inline src/monster.h]: PR looks fined. But why this change? Does it need to be ordered?

Also: can't you remove the old `mattack::leap` function and  convert the JSON to use the new function? To reduce code duplication and so.

Comment (Coolthulhu) [inline src/monster.h]: Forgot to explain this.

It's not directly related to the whole PR, but from what I could find, `unordered_map` needs to hash the entire key. In the case of string keys, that is roughly O(len(k)) complexity, where k is the key.
`map` only needs to read O(log n) symbols.

Since `special_attacks` is usually a very small container, it is highly likely that hashing will take more time than binary tree search.

Comment (kevingranade) [inline src/monster.h]: That's generally the case, and I'd also say that std::map is a bit more robust and space efficient in general, so it makes for a better default.
